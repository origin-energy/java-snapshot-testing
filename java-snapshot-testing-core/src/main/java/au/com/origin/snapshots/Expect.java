package au.com.origin.snapshots;

import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RequiredArgsConstructor
public class Expect {
  private final SnapshotVerifier snapshotVerifier;
  private final Method testMethod;

  private SnapshotSerializer snapshotSerializer;
  private SnapshotComparator snapshotComparator;
  private List<SnapshotReporter> snapshotReporters;
  private String scenario;

  private final Map<String, String> headers = new HashMap<>();

  public static Expect of(SnapshotVerifier snapshotVerifier, Method method) {
    return new Expect(snapshotVerifier, method);
  }

  /**
   * Make an assertion on the given input parameters against what already exists
   *
   * @param firstObject first snapshot object
   * @param objects     other snapshot objects
   */
  public void toMatchSnapshot(Object firstObject, Object... objects) {
    SnapshotContext snapshotContext = snapshotVerifier.expectCondition(testMethod, firstObject, objects);
    if (snapshotSerializer != null) {
      snapshotContext.setSnapshotSerializer(snapshotSerializer);
    }
    if (snapshotComparator != null) {
      snapshotContext.setSnapshotComparator(snapshotComparator);
    }
    if (snapshotReporters != null) {
      snapshotContext.setSnapshotReporters(snapshotReporters);
    }
    if (scenario != null) {
      snapshotContext.setScenario(scenario);
    }
    snapshotContext.header.putAll(headers);

    snapshotContext.toMatchSnapshot();
  }


  /**
   * Normally a snapshot can be applied only once to a test method.
   * <p>
   * For Parameterized tests where the same method is executed multiple times you can supply
   * the scenario() to overcome this restriction.  Ensure each scenario is unique.
   *
   * @param scenario - unique scenario description
   * @return Snapshot
   */
  public Expect scenario(String scenario) {
    this.scenario = scenario;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param serializer your custom serializer
   * @return Snapshot
   */
  public Expect serializer(SnapshotSerializer serializer) {
    this.snapshotSerializer = serializer;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param name - the {name} attribute serializer.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect serializer(String name) {
    this.snapshotSerializer = SnapshotProperties.getInstance("serializer." + name);
    return this;
  }

  /**
   * Apply a custom comparator for this snapshot
   *
   * @param comparator your custom comparator
   * @return Snapshot
   */
  public Expect comparator(SnapshotComparator comparator) {
    this.snapshotComparator = comparator;
    return this;
  }

  /**
   * Apply a custom comparator for this snapshot
   *
   * @param name the {name} attribute comparator.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect comparator(String name) {
    this.snapshotComparator = SnapshotProperties.getInstance("comparator." + name);
    return this;
  }

  /**
   * Apply a list of custom reporters for this snapshot
   * This will replace the default reporters defined in the config
   *
   * @param reporters your custom reporters
   * @return Snapshot
   */
  public Expect reporters(SnapshotReporter... reporters) {
    this.snapshotReporters = Arrays.asList(reporters);
    return this;
  }

  /**
   * Apply a list of custom reporters for this snapshot
   * This will replace the default reporters defined in the config
   *
   * @param name the {name} attribute reporters.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect reporters(String name) {
    this.snapshotReporters = SnapshotProperties.getInstances("reporters." + name);
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot.
   *
   * @param serializer your custom serializer
   * @return this
   * @see au.com.origin.snapshots.serializers.SnapshotSerializer
   * <p>
   * Example implementations
   * @see au.com.origin.snapshots.serializers.ToStringSnapshotSerializer
   * @see au.com.origin.snapshots.serializers.Base64SnapshotSerializer
   */
  @SneakyThrows
  public Expect serializer(Class<? extends SnapshotSerializer> serializer) {
    this.snapshotSerializer = serializer.getConstructor().newInstance();
    return this;
  }

  /**
   * Add anything you like to the snapshot header.
   *
   * These custom headers can be used in serializers, comparators or reporters to change
   * how they behave.
   *
   * @param key key
   * @param value value
   * @return Expect
   */
  public Expect header(String key, String value) {
    headers.put(key, value);
    return this;
  }

}
