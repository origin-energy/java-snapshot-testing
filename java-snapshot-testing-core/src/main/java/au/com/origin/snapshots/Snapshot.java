package au.com.origin.snapshots;

import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.exceptions.SnapshotMatchException;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Slf4j
public class Snapshot {

  private final SnapshotConfig snapshotConfig;
  private final SnapshotFile snapshotFile;
  private final Class<?> testClass;
  private final String snapshotName;
  private final Object[] current;
  private final boolean isCI;

  @Setter
  private SnapshotSerializer snapshotSerializer;
  @Setter
  private SnapshotComparator snapshotComparator;
  @Setter
  private List<SnapshotReporter> snapshotReporters;
  @Setter
  private String scenario;

  Snapshot(
      SnapshotConfig snapshotConfig,
      SnapshotFile snapshotFile,
      Class<?> testClass,
      String snapshotName,
      Object... current) {
    this.snapshotConfig = snapshotConfig;
    this.snapshotFile = snapshotFile;
    this.testClass = testClass;
    this.snapshotName = snapshotName;
    this.current = current;

    this.isCI = snapshotConfig.isCI();
    this.snapshotSerializer = snapshotConfig.getSerializer();
    this.snapshotComparator = snapshotConfig.getComparator();
    this.snapshotReporters = snapshotConfig.getReporters();
    this.scenario = null;
  }


  public void toMatchSnapshot() {

    Set<String> rawSnapshots = snapshotFile.getRawSnapshots();

    String rawSnapshot = getRawSnapshot(rawSnapshots);

    String currentObject = takeSnapshot();

    if (rawSnapshot != null && shouldUpdateSnapshot()) {
      snapshotFile.getRawSnapshots().remove(rawSnapshot);
      rawSnapshot = null;
    }

    if (rawSnapshot != null) {
      // Match existing Snapshot
      if (!snapshotComparator.matches(getSnapshotPath(), rawSnapshot, currentObject)) {
        snapshotFile.createDebugFile(currentObject.trim());

        List<SnapshotReporter> reporters = snapshotReporters
            .stream()
            .filter(reporter -> reporter.supportsFormat(snapshotSerializer.getOutputFormat()))
            .collect(Collectors.toList());

        if (reporters.isEmpty()) {
          String comparator = snapshotComparator.getClass().getSimpleName();
          throw new IllegalStateException("No compatible reporters found for comparator " + comparator);
        }

        List<Throwable> errors = new ArrayList<>();

        for (SnapshotReporter reporter : reporters) {
          try {
            reporter.report(getSnapshotPath(), rawSnapshot, currentObject);
          } catch (Throwable t) {
            errors.add(t);
          }
        }

        if (!errors.isEmpty()) {
          throw new SnapshotMatchException("Error(s) matching snapshot(s)", errors);
        }
      }
    } else {
      if (this.isCI) {
        log.error("We detected you are running on a CI Server - if this is incorrect please override the isCI() method in SnapshotConfig");
        throw new SnapshotMatchException("Snapshot [" + getSnapshotPath() + "] not found. Has this snapshot been committed ?");
      } else {
        log.warn("We detected you are running on a developer machine - if this is incorrect please override the isCI() method in SnapshotConfig");
        // Create New Snapshot
        snapshotFile.push(currentObject);
      }
    }
    snapshotFile.deleteDebugFile();
  }

  private boolean shouldUpdateSnapshot() {
    if (snapshotConfig.updateSnapshot().isPresent()) {
      return getSnapshotPath().contains(snapshotConfig.updateSnapshot().get());
    } else {
      return false;
    }
  }

  private String getRawSnapshot(Collection<String> rawSnapshots) {
    for (String rawSnapshot : rawSnapshots) {
      if (rawSnapshot.contains(getSnapshotPath())) {
        return rawSnapshot;
      }
    }
    return null;
  }

  private String takeSnapshot() {
    return getSnapshotPath() + snapshotSerializer.apply(current);
  }

  String getSnapshotPath() {
    String scenarioFormat = scenario == null ? "" : "[" + scenario + "]";
    return snapshotName + scenarioFormat + "=";
  }
}
