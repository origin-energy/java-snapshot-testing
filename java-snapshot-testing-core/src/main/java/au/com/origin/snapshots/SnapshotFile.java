package au.com.origin.snapshots;

import lombok.Getter;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.Collections;
import java.util.Set;
import java.util.TreeSet;
import java.util.function.BiFunction;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Slf4j
public class SnapshotFile {

  public static final String SPLIT_STRING = "\n\n\n";

  private final String fileName;
  private final Class<?> testClass;
  private final BiFunction<Class<?>, String, String> onSaveSnapshotFile;
  @Getter
  private Set<String> rawSnapshots;

  SnapshotFile(String srcDirPath, String fileName, Class<?> testClass, BiFunction<Class<?>, String, String> onSaveSnapshotFile) throws IOException {
    this.testClass = testClass;
    this.onSaveSnapshotFile = onSaveSnapshotFile;
    this.fileName = srcDirPath + File.separator + fileName;
    log.info("Snapshot File: " + this.fileName);

    StringBuilder fileContent = new StringBuilder();

    try (BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream(this.fileName), StandardCharsets.UTF_8))) {

      String sCurrentLine;

      while ((sCurrentLine = br.readLine()) != null) {
        fileContent.append(sCurrentLine + "\n");
      }

      String fileText = fileContent.toString();
      if (!"".equals(fileText.trim())) {
        rawSnapshots =
            Collections.synchronizedSortedSet(
                Stream.of(fileContent.toString().split(SPLIT_STRING))
                    .map(String::trim)
                    .collect(Collectors.toCollection(TreeSet::new)));
      } else {
        rawSnapshots = Collections.synchronizedSortedSet(new TreeSet<>());
      }
    } catch (IOException e) {
      rawSnapshots = Collections.synchronizedSortedSet(new TreeSet<>());
    }
  }

  private String getDebugFilename() {
    return this.fileName + ".debug";
  }

  public File createDebugFile(String snapshot) {
    File file = null;
    try {
      file = new File(getDebugFilename());
      file.getParentFile().mkdirs();
      file.createNewFile();

      try (FileOutputStream fileStream = new FileOutputStream(file, false)) {
        fileStream.write(snapshot.getBytes(StandardCharsets.UTF_8));
      } catch (IOException e) {
        e.printStackTrace();
      }
    } catch (IOException e) {
      e.printStackTrace();
    }

    return file;
  }

  @SneakyThrows
  public void deleteDebugFile() {
    Files.deleteIfExists(Paths.get(getDebugFilename()));
  }

  @SneakyThrows
  public void delete() {
    Files.deleteIfExists(Paths.get(this.fileName));
  }

  @SneakyThrows
  public synchronized File createFileIfNotExists() {
    Path path = Paths.get(this.fileName);
    if (!Files.exists(path)) {
      Files.createDirectories(path.getParent());
      Files.createFile(path);
    }
    return path.toFile();
  }

  public void push(String snapshot) {
    rawSnapshots.add(snapshot);

    File file = createFileIfNotExists();

    try (FileOutputStream fileStream = new FileOutputStream(file, false)) {
      byte[] myBytes = String.join(SPLIT_STRING, rawSnapshots).getBytes(StandardCharsets.UTF_8);
      fileStream.write(myBytes);
    } catch (IOException e) {
      e.printStackTrace();
    }
  }

  @SneakyThrows
  public void cleanup() {
    Path path = Paths.get(this.fileName);
    if (Files.exists(path)) {
      if (Files.size(path) == 0) {
        deleteDebugFile();
        delete();
      } else {
        String content = new String(Files.readAllBytes(Paths.get(this.fileName)), StandardCharsets.UTF_8);
        String modified = onSaveSnapshotFile.apply(testClass, content);
        Files.write(path, modified.getBytes(StandardCharsets.UTF_8), StandardOpenOption.TRUNCATE_EXISTING);
      }
    }
  }
}
