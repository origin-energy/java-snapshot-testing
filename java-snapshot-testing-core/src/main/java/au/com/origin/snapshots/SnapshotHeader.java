package au.com.origin.snapshots;

import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@RequiredArgsConstructor
public class SnapshotHeader extends HashMap<String, String> {

    //
    // Manual JSON serialization/deserialization as I don't want to
    // include another dependency for it
    //
    @SneakyThrows
    public String toJson() {
        StringBuilder b = new StringBuilder();
        b.append("{\n");
        final int lastIndex = this.size();
        int currIndex = 0;
        for (Map.Entry entry : this.entrySet()) {
            currIndex++;
            String format = currIndex == lastIndex ? "  \"%s\": \"%s\"\n" : "  \"%s\": \"%s\",\n";
            b.append(String.format(format, entry.getKey(), entry.getValue()));
        }
        b.append("}");
        return b.toString();
    }

    @SneakyThrows
    public static SnapshotHeader fromJson(String json) {
        SnapshotHeader snapshotHeader = new SnapshotHeader();

        if (json == null) {
            return snapshotHeader;
        }

        String regex = "\\\"(?<key>.*)\\\": \\\"(?<value>.*)\\\"";
        Pattern p = Pattern.compile(regex);
        Matcher m = p.matcher(json);
        while (m.find()) {
            snapshotHeader.put(m.group("key"), m.group("value"));
        }
        return snapshotHeader;
    }
}
