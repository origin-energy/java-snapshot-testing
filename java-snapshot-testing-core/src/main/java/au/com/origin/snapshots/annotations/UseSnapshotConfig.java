package au.com.origin.snapshots.annotations;

import au.com.origin.snapshots.config.SnapshotConfig;

import java.lang.annotation.*;

@Documented
@Target({ElementType.TYPE})
@Inherited
@Retention(RetentionPolicy.RUNTIME)
public @interface UseSnapshotConfig {
  Class<? extends SnapshotConfig> value();
}
