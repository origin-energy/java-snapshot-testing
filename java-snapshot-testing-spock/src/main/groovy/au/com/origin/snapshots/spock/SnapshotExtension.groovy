package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotVerifier
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfigInjector
import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo

class SnapshotExtension extends AbstractAnnotationDrivenExtension<EnableSnapshots> implements SnapshotConfigInjector {

    SnapshotVerifier snapshotVerifier;

    void visitSpecAnnotation(EnableSnapshots annotation, SpecInfo spec) {
        this.snapshotVerifier = new SnapshotVerifier(getSnapshotConfig(), spec.reflection, false)
    }

    void visitSpec(SpecInfo spec) {
        spec.allFeatures.featureMethod*.addInterceptor(new SnapshotMethodInterceptor(snapshotVerifier))
    }



    @Override
    SnapshotConfig getSnapshotConfig() {
        return new PropertyResolvingSnapshotConfig()
    }
}