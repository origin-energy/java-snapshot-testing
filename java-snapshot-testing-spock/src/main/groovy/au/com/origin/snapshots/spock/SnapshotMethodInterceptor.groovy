package au.com.origin.snapshots.spock

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.SnapshotVerifier
import org.spockframework.runtime.extension.AbstractMethodInterceptor
import org.spockframework.runtime.extension.IMethodInterceptor
import org.spockframework.runtime.extension.IMethodInvocation

import java.lang.reflect.Method

// Based on this issue: https://github.com/spockframework/spock/issues/652
class SnapshotMethodInterceptor extends AbstractMethodInterceptor {

    private final SnapshotVerifier snapshotVerifier;

    SnapshotMethodInterceptor(SnapshotVerifier snapshotVerifier) {
        this.snapshotVerifier = snapshotVerifier
    }

    @Override
    void interceptFeatureMethod(IMethodInvocation invocation) throws Throwable {
        updateInstanceVariable(invocation.instance, invocation.feature.featureMethod.reflection)

        def parameterCount = invocation.method.reflection.parameterCount
        if (parameterCount > invocation.arguments.length) {
            def newArguments = new Object[parameterCount]
            System.arraycopy invocation.arguments, 0, newArguments, 0, invocation.arguments.length
            invocation.arguments = newArguments
        }
        invocation.method.reflection.parameterTypes.eachWithIndex { type, i ->
            if (Expect.class == type) {
                invocation.arguments[i] = new Expect(snapshotVerifier, invocation.feature.featureMethod.reflection)
            }
        }
        invocation.proceed()
    }

    private void updateInstanceVariable(Object testInstance, Method testMethod) {
        testInstance.class.declaredFields
            .find { it.getType() == Expect.class }
            ?.with {
                Expect expect = Expect.of(snapshotVerifier, testMethod)
                it.setAccessible(true)
                it.set(testInstance, expect)
            }
    }

    @Override
    void interceptCleanupSpecMethod(IMethodInvocation invocation) throws Throwable {
        this.snapshotVerifier.validateSnapshots();
    }
}
