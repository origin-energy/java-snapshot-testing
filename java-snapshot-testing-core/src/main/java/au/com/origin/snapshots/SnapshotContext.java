package au.com.origin.snapshots;

import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.config.SnapshotConfig;
import au.com.origin.snapshots.exceptions.SnapshotMatchException;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Slf4j
public class SnapshotContext {

  private final SnapshotConfig snapshotConfig;
  private final SnapshotFile snapshotFile;

  @Getter
  final Class<?> testClass;

  @Getter
  final Method testMethod;

  final Object[] current;
  private final boolean isCI;

  @Setter
  private SnapshotSerializer snapshotSerializer;
  @Setter
  private SnapshotComparator snapshotComparator;
  @Setter
  private List<SnapshotReporter> snapshotReporters;

  @Setter
  @Getter
  String scenario;

  @Getter
  SnapshotHeader header = new SnapshotHeader();

  SnapshotContext(
      SnapshotConfig snapshotConfig,
      SnapshotFile snapshotFile,
      Class<?> testClass,
      Method testMethod,
      Object... current) {
    this.snapshotConfig = snapshotConfig;
    this.snapshotFile = snapshotFile;
    this.testClass = testClass;
    this.testMethod = testMethod;
    this.current = current;

    this.isCI = snapshotConfig.isCI();
    this.snapshotSerializer = snapshotConfig.getSerializer();
    this.snapshotComparator = snapshotConfig.getComparator();
    this.snapshotReporters = snapshotConfig.getReporters();
    this.scenario = null;
  }


  public void toMatchSnapshot() {

    Set<Snapshot> rawSnapshots = snapshotFile.getSnapshots();
    Snapshot previousSnapshot = getRawSnapshot(rawSnapshots);
    Snapshot currentSnapshot = takeSnapshot();

    if (previousSnapshot != null && shouldUpdateSnapshot()) {
      snapshotFile.getSnapshots().remove(previousSnapshot);
      previousSnapshot = null;
    }

    if (previousSnapshot != null) {
      // Match existing Snapshot
      if (!snapshotComparator.matches(previousSnapshot, currentSnapshot)) {
        snapshotFile.createDebugFile(currentSnapshot);

        List<SnapshotReporter> reporters = snapshotReporters
            .stream()
            .filter(reporter -> reporter.supportsFormat(snapshotSerializer.getOutputFormat()))
            .collect(Collectors.toList());

        if (reporters.isEmpty()) {
          String comparator = snapshotComparator.getClass().getSimpleName();
          throw new IllegalStateException("No compatible reporters found for comparator " + comparator);
        }

        List<Throwable> errors = new ArrayList<>();

        for (SnapshotReporter reporter : reporters) {
          try {
            reporter.report(previousSnapshot, currentSnapshot);
          } catch (Throwable t) {
            errors.add(t);
          }
        }

        if (!errors.isEmpty()) {
          throw new SnapshotMatchException("Error(s) matching snapshot(s)", errors);
        }
      }
    } else {
      if (this.isCI) {
        log.error("We detected you are running on a CI Server - if this is incorrect please override the isCI() method in SnapshotConfig");
        throw new SnapshotMatchException("Snapshot [" + resolveSnapshotIdentifier() + "] not found. Has this snapshot been committed ?");
      } else {
        log.warn("We detected you are running on a developer machine - if this is incorrect please override the isCI() method in SnapshotConfig");
        // Create New Snapshot
        snapshotFile.push(currentSnapshot);
      }
    }
    snapshotFile.deleteDebugFile();
  }

  private boolean shouldUpdateSnapshot() {
    if (snapshotConfig.updateSnapshot().isPresent()) {
      return resolveSnapshotIdentifier().contains(snapshotConfig.updateSnapshot().get());
    } else {
      return false;
    }
  }

  private Snapshot getRawSnapshot(Collection<Snapshot> rawSnapshots) {
    for (Snapshot rawSnapshot : rawSnapshots) {
      if (rawSnapshot.getIdentifier().equals(resolveSnapshotIdentifier())) {
        return rawSnapshot;
      }
    }
    return null;
  }

  private Snapshot takeSnapshot() {
    SnapshotSerializerContext sg = SnapshotSerializerContext.from(this);
    return snapshotSerializer.apply(current, sg);
  }

  String resolveSnapshotIdentifier() {
    String scenarioFormat = scenario == null ? "" : "[" + scenario + "]";
    SnapshotName snapshotName = testMethod.getAnnotation(SnapshotName.class);
    String pathFormat = snapshotName == null ?
        testClass.getName() + "." + testMethod.getName() :
        snapshotName.value();
    return pathFormat + scenarioFormat;
  }
}
