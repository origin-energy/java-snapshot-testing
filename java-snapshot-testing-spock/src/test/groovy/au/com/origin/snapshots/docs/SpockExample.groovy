package au.com.origin.snapshots.docs

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.annotations.SnapshotName
import au.com.origin.snapshots.spock.EnableSnapshots
import spock.lang.Specification

// Ensure you enable snapshot testing support
@EnableSnapshots
class SpockExample extends Specification {

    // Option 1: inject Expect as an instance variable
    private Expect expect

    // With spock tests you should always use @SnapshotName - otherwise they become coupled to test order
    @SnapshotName("should_use_extension")
    def "Should use extension"() {
        when:
        expect.toMatchSnapshot("Hello World")

        then:
        true
    }

    @SnapshotName("should_use_extension_as_method_argument")
    // Option 2: inject Expect into the method signature
    def "Should use extension as method argument"(Expect expect) {
        when:
        expect.toMatchSnapshot("Hello World")

        then:
        true
    }
}
