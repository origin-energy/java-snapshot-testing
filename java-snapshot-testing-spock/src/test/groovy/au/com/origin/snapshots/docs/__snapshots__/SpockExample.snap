should_use_extension=[
Hello World
]


should_use_extension_as_method_argument=[
Hello World
]
